src/wasm-interpreter.h

```diff
@@ -2228,6 +2235,119 @@ class ModuleInstance
+
+  // Special exception indicating a flow not evaluating to a constant
+  struct NonconstantException {
+  }; // TODO: use a flow with a special name, as this is likely very slow
+
+  ContextAwareExpressionRunner(Module* module,
+                               Flags flags,
+                               Index maxDepth,
+                               GetValues* getValues = nullptr);
+
+  // Gets the module this runner belongs to.
+  Module* getModule();
+
+  // Sets a known local value to use. Returns `true` if the value is actually
+  // constant.
+  bool setLocalValue(Index index, Literals& values);
+
+  // Sets a known local value to use. Order matters if expressions have side
+  // effects. Returns `true` if the expression actually evaluates to a constant.
+  bool setLocalValue(Index index, Expression* expr);
+
+  // Sets a known global value to use. Returns `true` if the value is actually
+  // constant.
+  bool setGlobalValue(Name name, Literals& values);
+
+  // Sets a known global value to use. Order matters if expressions have side
+  // effects. Returns `true` if the expression actually evaluates to a constant.
+  bool setGlobalValue(Name name, Expression* expr);
+
+  Flow visitLoop(Loop* curr);
+  Flow visitCall(Call* curr);
+  Flow visitCallIndirect(CallIndirect* curr);
+  Flow visitLocalGet(LocalGet* curr);
+  Flow visitLocalSet(LocalSet* curr);
+  Flow visitGlobalGet(GlobalGet* curr);
+  Flow visitGlobalSet(GlobalSet* curr);
+  Flow visitLoad(Load* curr);
+  Flow visitStore(Store* curr);
+  Flow visitAtomicRMW(AtomicRMW* curr);
+  Flow visitAtomicCmpxchg(AtomicCmpxchg* curr);
+  Flow visitAtomicWait(AtomicWait* curr);
+  Flow visitAtomicNotify(AtomicNotify* curr);
+  Flow visitSIMDLoad(SIMDLoad* curr);
+  Flow visitMemoryInit(MemoryInit* curr);
+  Flow visitDataDrop(DataDrop* curr);
+  Flow visitMemoryCopy(MemoryCopy* curr);
+  Flow visitMemoryFill(MemoryFill* curr);
+  Flow visitHost(Host* curr);
+  Flow visitTry(Try* curr);
+  Flow visitThrow(Throw* curr);
+  Flow visitRethrow(Rethrow* curr);
+  Flow visitBrOnExn(BrOnExn* curr);
+  Flow visitPush(Push* curr);
+  Flow visitPop(Pop* curr);
+  void trap(const char* why) override;
+
+private:
+  // Module this runner belongs to.
+  Module* module;
+
+  // Map of local indexes to values set in the context of this runner
+  std::unordered_map<Index, Literals> localValues;
+
+  // Map of global names to values set in the context of this runner
+  std::unordered_map<Name, Literals> globalValues;
+
+  // Optional reference to a map of gets to constant values, for example where a
+  // pass did already compute these. Not applicable in C-API usage and possibly
+  // large, hence a reference to the respective map (must not be mutated).
+  GetValues* getValues;
+
+  // Flags indicating special requirements, for example whether we are just
+  // evaluating (default), also going to replace the expression afterwards or
+  // executing in a function-parallel scenario.
```

Comment:
this comment is duplicated above - one of the places would be enough I think.

Reply:
Shortened the comment, referring to `FlagValues`. So we still have intellisense on it, but less duplicate text.